### PR TITLE
feat(mobile): Library page rework (album sorting, favorites)

### PR DESCRIPTION
**Pull request by matthinc**
Adds sorting for albums and favorites management on mobile.

### CODE REVIEW

File: mobile/lib/modules/favorite/providers/favorite_provider.dart [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import 'package:hooks_riverpod/hooks_riverpod.dart';
+import 'package:immich_mobile/shared/providers/asset.provider.dart';
+import 'package:openapi/api.dart';
+
+class FavoriteSelectionNotifier extends StateNotifier<Set<String>> {
```
Comment (matthinc) [inline mobile/lib/modules/favorite/providers/favorite_provider.dart:5]: This was necessary for performance reasons. Updating the assetProvider all the time was too slow.
This solution also works in albums.

Comment (martyfuhry) [reply mobile/lib/modules/favorite/providers/favorite_provider.dart:5]: Looks good!

File: mobile/lib/modules/favorite/views/favorites_page.dart [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+          style: TextStyle(fontSize: 16, fontWeight: FontWeight.bold),
+        ).tr(),
+      );
+    }
+
+    Widget buildImageGrid() {
+      final appSettingService = ref.watch(appSettingsServiceProvider);
+
+      if (ref.watch(favoriteAssetProvider).isNotEmpty) {
+        return SliverPadding(
```
Comment (martyfuhry) [inline mobile/lib/modules/favorite/views/favorites_page.dart:34]: Any reason why you are using a `SliverGrid` and not just `GridView`? `Slivers` are much more extensible but they can be more difficult to work.

**Review by martyfuhry — COMMENTED** [review]
My (thin) phone app bar is getting crowded. We will need a new issue to fix this as more items are added. The favorites icon is already crowding out my back button.
image

I wouldn't say it's critical to fix for this PR, but this has introduced the issue for me.

**Comment by martyfuhry** [comment]
One more thing, you should add `Hero` animations to the Favorites grid / detail view! It would animate them very nicely as you select and close them.

File: mobile/lib/modules/album/ui/album_viewer_thumbnail.dart [hunk 3]
```diff
@@ -141,6 +154,8 @@ class AlbumViewerThumbnail extends HookConsumerWidget {
       child: Stack(
         children: [
           buildThumbnailImage(),
+          if (ref.watch(favoriteProvider).contains(asset.id))
```
[the remark this answers is not in the document]
Comment (jrasm91) [reply mobile/lib/modules/album/ui/album_viewer_thumbnail.dart:157]: I'll just say - the way google handles this is the horizontal overflow menu (when clicked), opens the bottom sheet, which has more actions. I _really dont_ like this. The page has actions in the appbar (cast, favorite), more actions in the bottom bar (share, edit, lens, delete), and more at the top of the bottom sheet (add to album, archive, delete from device, order photo, move to locked folder, use as slideshow, print, help/feedback). It feels really confusing to me trying to guess/remember which action is where. So, agree we should (eventually, soonish?) figure out a way to deal with new icons/actions, but I also don't love the gphotos approach.

[the remark this answers is not in the document]
Comment (martyfuhry) [reply mobile/lib/modules/album/ui/album_viewer_thumbnail.dart:157]: Agreed. I think you meant to respond to my comment with the screenshot, right? I meant in this suggestion only that you should add an additional variable name `isFavorite` and move the boolean logic to above this declaration, not remove the icon. I hope that's more clear!

[the remark this answers is not in the document]
Comment (jrasm91) [reply mobile/lib/modules/album/ui/album_viewer_thumbnail.dart:157]: Oh yes, in the main "conversations" page it shows up as nested comment under the main "review" comment, but yes that response was meant with regards to the overflowing icon comment.

File: mobile/lib/modules/favorite/views/favorites_page.dart (hunk 2 above)
Comment (matthinc) [reply mobile/lib/modules/favorite/views/favorites_page.dart:34]: I'd close this for now as we agreed on refactoring all asset grids soon @martyfuhry .

**Comment by martyfuhry** [comment]
Looks like there are some merge conflicts. Other than that, this looks good to me!

I found one odd bug quickly tapping on the "Favorite" star on an asset. If you tap too quickly, the toggling seems to lag. When you stop quickly tapping, the star will continue to toggle. My guess is that the UI state is a little too tightly coupled with the business logic updating the asset's value.

I'll happily approve even with the known bug, but I wanted to make sure it was recorded somewhere!

**Review by martyfuhry — APPROVED** [review]

**Comment by martyfuhry** [comment]
More conflicts, looks like.

**Pull request merged**